### PR TITLE
Reimplement nodes

### PR DESCRIPTION
**Pull request by paolobrasolin**

### CODE REVIEW

**Comment by jonsterling** [comment]
OMG! I’m super excited to try this out!!

**Comment by jonsterling** [comment]
Holy crap, this really does seem to work! I love that you deleted a lot of terrible code that I wasn't happy with. I'll review this more carefully tomorrow.

Thank you so much for your work!

File: _plugins/import_tag.rb [hunk 1]
```diff
@@ -1,30 +1,11 @@
-# This code is roughly based on https://github.com/koraktor/jekyll/tree/import-tag
-#
 # frozen_string_literal: true
 
+# TODO: this still exists only to avoid a crash with jekyll-feed, and we should find a better solution. This can mean ensuring jekyll-feed skips nodes, or choosind a different syntax for imports, or...
```
Comment (jonsterling) [inline _plugins/import_tag.rb:3]: My understanding is that the import is now in your version implemented *not* as a Liquid tag; therefore, I think it makes sense to consider a different syntax if that would help.

File: _plugins/gnarly.rb [hunk 2]
```diff
@@ -0,0 +1,113 @@
+require 'tsort'
+
+module Gnarly
```
Comment (jonsterling) [inline _plugins/gnarly.rb]: Shall we call this module something different? 😜

File: _nodes/0006.md [hunk 3]
```diff
@@ -1,6 +1,5 @@
 ---
 title: "Example: the family fibration"
-layout: page
```
Comment (jonsterling) [inline _nodes/0006.md:3]: Thank god I don't have to keep typing this line LOL.

One question: is it possible to make the `macrolib` be inherited from the root?

File: _plugins/gnarly.rb [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    include TSort
+
+    alias tsort_each_node each_key
+
+    def tsort_each_child(node, &block)
+      fetch(node).each(&block)
+    end
+  end
+
+  def self.render_header(resource, level)
```
Comment (jonsterling) [inline _plugins/gnarly.rb]: For my own edification, what does `def self.(...)` do in Ruby?

Comment (jonsterling) [reply _plugins/gnarly.rb]: Ah never mind, I get it now!

**Review by jonsterling — APPROVED** [review]
@paolobrasolin This is totally fantastic! The code and architecture are a zillion times better than what I did, and much more likely to scale up to interesting and useful features going forward. Really nice work. I am very happy with merging this now, let me know if you'd like to get any more commits in before I do so; also let me know if you'd like me to just give you commit-access to this repo.

I agree with you that it would make sense to consider breaking this functionality out into a package soon that people can depend on. To answer your question about what additional features of Gerby would be nice to have in the future:

1. Being able to go "next" and "previous" would be great.
2. Search of some kind would be cool (but sounds hard, not a huge priority).

In general however my preference is to keep the footprint fairly small. Simple functionality that works well will lead to an explosion of high quality lecture notes and reference materials, which I'm very excited about.

File: _plugins/import_tag.rb (hunk 1 above)
Comment (paolobrasolin) [reply _plugins/import_tag.rb:3]: I found a better option in setting out hook with a higher priority than `jekyll-feed`'s. I have to test it yet, but it should work.

Still, using the syntax of a liquid tag might be misleading.
Maybe something similar to a "preprocessor directive" like `#include 000F` would be better.

Comment (jonsterling) [reply _plugins/import_tag.rb:3]: Gotcha! My only thought is that `#include` might be a bit confusing because of Markdown. What about `@include 000F`? I think this doesn't conflict with anything.

File: _plugins/gnarly.rb (hunk 2 above)
Comment (paolobrasolin) [reply _plugins/gnarly.rb]: Haha. Yeah, it was processing a lot of tree nodes, so "gnarly wood" came to mind.

Maybe `sheafy` would be nicer, as it's a simpler version of `gerby`. How do you feel about that?

File: _nodes/0006.md (hunk 3 above)
Comment (paolobrasolin) [reply _nodes/0006.md:3]: Yes! You can set front matter defaults by file path and collection type -- see [details here](https://jekyllrb.com/docs/configuration/front-matter-defaults/).

File: _plugins/gnarly.rb (hunk 2 above)
Comment (jonsterling) [reply _plugins/gnarly.rb]: Oh I thought you were just calling it that because it was painful to write 😆

File: _nodes/0006.md (hunk 3 above)
Comment (jonsterling) [reply _nodes/0006.md:3]: Ah but I think this doesn't exactly do what I need --- In general I may have multiple root "lectures" that import different nodes; each lecture may use a different macro library. Do you see how to accommodate this usage, or do you have a suggestion for another way for me to organize my notes?

Comment (jonsterling) [reply _nodes/0006.md:3]: (It is also OK to say "Don't have multiple macro libraries!!". I could make peace with that potentially.)

**Comment by paolobrasolin** [comment]
> @paolobrasolin This is totally fantastic! The code and architecture are a zillion times better than what I did, and much more likely to scale up to interesting and useful features going forward. Really nice work.

I'm glad you like this! :smile: 

> I am very happy with merging this now, let me know if you'd like to get any more commits in before I do so;

I just renamed `Gnarly` to `Sheafy`, so I think you can go on with the merge and use this.
We can think of the details later. :+1: 

> also let me know if you'd like me to just give you commit-access to this repo.

Sounds great! If that doesn't bother you, It'll make keeping up simpler for me.

> I agree with you that it would make sense to consider breaking this functionality out into a package soon that people can depend on.

`jekyll-sheafy` it is then! :smiley: 

> To answer your question about what additional features of Gerby would be nice to have in the future:
> 
> 1. Being able to go "next" and "previous" would be great.
> 2. Search of some kind would be cool (but sounds hard, not a huge priority).
> 
> In general however my preference is to keep the footprint fairly small.

I agree with you. I'll sleep on those two ideas and see if there are simple solutions. 

> Simple functionality that works well will lead to an explosion of high quality lecture notes and reference materials, which I'm very excited about.

**Exactly!** Let's make that happen! :rocket:

File: _nodes/0006.md (hunk 3 above)
Comment (paolobrasolin) [reply _nodes/0006.md:3]: Oh ok, I misunderstood your question.

Yes, we can simply use [this iteration](https://github.com/jonsterling/math/blob/d42ab9f8b56b243414c431312cdd798141f0280e/_plugins/sheafy.rb#L90) to push front matter values down from roots to leaves in a single pass.

The only issue I see with that is with children nodes belonging to multiple parents; that's not happening in this use case, but it's also not explicitly forbidden and might cause unexpected behaviour.

Comment (jonsterling) [reply _nodes/0006.md:3]: Cool, thanks. My take is that both scenarios are badly behaved --- in particular, a node that depends on `macrolib1` cannot be rendered under any circumstances alongside a node that depends on `macrolib2`, because the KaTeX architecture imports macros for the entire page and not for specific regions of the page. (Maybe this could be addressed but I think it is asking for suffering, and not likely to satisfy many interesting use-cases). This is all to say that I think we should expect bad behavior in this case.

My original plan was to have one "math website" with a lot of different lectures/mini-books in it. Each of these would naturally require different macro-libraries. But it may be that this style of working is bad, and that one should instead imagine having one "sheafy website" per mini-book. In that case it would be natural to fix a global macro library for the entire thing.  This might be simpler for people to understand and probably would not rule out many real applications. 

What do you think?

Comment (paolobrasolin) [reply _nodes/0006.md:3]: Ok, let me think aloud a bit:

- a node must be renderable as a standalone;
- a node must be renderable when embedded;
- a node must always yield the same result when rendered.

So basically embedding/rendering must compositionally well-behaved.
This greatly restricts what the factors altering rendering (or embedding) can do.

In particular, this implies that every node must always have the same macro set of any of its parents, and therefore of any of its ancestors.

It seems like that setting this property (macro set) only on ancestors and letting it dribble down (like I suggested in the previous message) should be the correct thing to do.

However, this breaks when a node is a descendant of two nodes which have no common ancestor: if the latter have different properties, an ambiguity happens.

Forbidding this situation is equivalent to asking that each connected component of the DAG is rooted, i.e. it has a base node which is an ancestor of every other one therefore fixing the property on the component.

We could even forbid reusing nodes, therefore making the DAG a **Directed Rooted Forest**.
Should we? I can think of three reasons:

- hyperlinking from the TOC to the anchors of different in-page instances of the same node would be annoying
- reasoning about trees would be easier for the user
- rooted trees are exactly what `gerby` produces from LaTeX documents

TL;DR: we should simply check the DAG is a **DRF** and let some selected properties dribble down the trees; this allows you to maintain multiple separate mini-books within the same Jekyll installation in a very natural way.

Comment (jonsterling) [reply _nodes/0006.md:3]: Thanks for this, sorry for the delay in getting back to you! I agree with your analysis of the situation, and come to the same conclusions as you. In particular, I think that it actually makes a lot of sense to enforce DRFs; this simplifies a lot of things while only ruling out what I would consider to be pathological use-cases of the tool. In general, while we should be in principle open to deviating from the Gerby model, Gerby also provides a wealth of experience about what is and is not useful. So I like and endorse your idea. Thanks!

Comment (paolobrasolin) [reply _nodes/0006.md:3]: Great! As you saw I prepared a nest at https://github.com/paolobrasolin/jekyll-sheafy
All topological checks are implemented in https://github.com/paolobrasolin/jekyll-sheafy/blob/main/lib/jekyll/sheafy/directed_graph.rb
There's a PR to use them in `math`: https://github.com/jonsterling/math/pull/13
